File: repository/src/main/java/com/mn/githubrepository/usecase/GetRepos.java
package com.mn.githubrepository.usecase;

import com.mn.githubrepository.api.GithubApi;
import com.mn.githubrepository.data.DataStore;
import com.mn.githubrepository.data.GithubRepo;

import io.reactivex.Flowable;

public class GetRepos {
    private final GithubApi githubApi;
    private final DataStore<GithubRepo> dataStore;
    private final String username;

    public GetRepos(GithubApi githubApi, DataStore<GithubRepo> dataStore, String username) {
        this.githubApi = githubApi;
        this.dataStore = dataStore;
        this.username = username;
    }

    public Flowable<Object> get(int page, int perPage) {
        //todo persist locally.
        //todo if the call fails and if the page is 0, read from local store.
        return githubApi.getRepos(username, page, perPage);
    }
}

File: repository/src/main/java/com/mn/githubrepository/usecase/GetRepos.kt
package com.mn.githubrepository.usecase

import com.mn.githubrepository.api.GithubApi
import com.mn.githubrepository.data.DataStore
import com.mn.githubrepository.data.GithubRepo
import com.mn.githubrepository.thread.ThreadTransformer
import io.reactivex.Flowable
import io.reactivex.Single
import org.reactivestreams.Publisher
import org.reactivestreams.Subscriber

class GetRepos(private val githubApi: GithubApi,
               private val dataStore: DataStore,
               private val username: String,
               private val threadTransformer: ThreadTransformer) {

    fun fetch(page: Int, perPage: Int): Flowable<List<GithubRepo>> {
        val localFlowable = Flowable.fromIterable(dataStore.getItems())
                .toList()
                .toFlowable()

        val remoteFlowable = githubApi.getRepos(username, page, perPage)
                .compose(threadTransformer.apply())
                .doOnNext { repos: List<GithubRepo>? ->
                    persistToDataStore(repos)
                }
        return if (page == 0) {
            Flowable.concat(localFlowable, remoteFlowable)
        } else {
            remoteFlowable
        }
    }

    private fun persistToDataStore(repos: List<GithubRepo>?) {
        repos?.forEach { githubRepo ->
            dataStore.addItem(githubRepo)
        }
    }
}
